contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -186,16 +209,179 @@ pub fn set_chain_config(
+        ensure!(
+            origin_chain == ChainNameRaw::from(message_direction.clone()),
+            Error::TokenDeployedFromNonOriginChain {
+                token_id: message.token_id(),
+                origin_chain,
+                chain: message_direction.clone().into(),
+            }
+        );
+
+        // Token chain info must already exist from previous deployment
+        ensure!(
+            token_chain_info.is_some(),
+            Error::TokenNotDeployed {
+                token_id: message.token_id(),
+                chain: message_direction.clone().into(),
+            }
+        );
+    } else {
+        // Token is being deployed for the first time
+        let token_config = TokenConfig {
+            origin_chain: ChainNameRaw::from(message_direction.clone()),
+        };
+
+        state::save_token_config(storage, &message.token_id(), &token_config)
+            .change_context(Error::State)?;
+
+        ensure!(
+            token_chain_info.is_none(),
+            Error::TokenAlreadyDeployed {
+                token_id: message.token_id(),
+                chain: message_direction.clone().into(),
+            }
+        );
+    }
+
+    Ok(())
+}
+
+/// Ensures that the token is not being redeployed to the same destination chain.
+fn token_redeployment_check(
+    message: &Message,
+    message_direction: &MessageDirection,
+    token_chain_info: &Option<TokenChainInfo>,
+) -> Result<(), Error> {
+    if matches!(
+        message,
+        Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. }
+    ) && matches!(message_direction, MessageDirection::To(_))
+        && token_chain_info.is_some()
+    {
+        bail!(Error::TokenAlreadyDeployed {
+            token_id: message.token_id(),
+            chain: message_direction.clone().into(),
+        });
+    }
+
+    Ok(())
+}
+
+/// Ensures that the token info is recorded on deployment.
+/// The token supply tracking is also enabled on deployment for trustless token deployments (i.e no minter set).
+/// Tokens that haven't been deployed yet cannot be transferred.
+fn token_deployment_check(
+    message: &Message,
+    message_direction: &MessageDirection,
+    token_chain_info: Option<TokenChainInfo>,
+) -> Result<TokenChainInfo, Error> {
+    if let Some(token_chain_info) = token_chain_info {
+        return Ok(token_chain_info);
+    }
+
+    let token_deployment_type = match message {
+        Message::DeployInterchainToken { minter: None, .. } => TokenDeploymentType::Trustless,
+        Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. } => {
+            TokenDeploymentType::CustomMinter
+        }
+        Message::InterchainTransfer { .. } => bail!(Error::TokenNotDeployed {
+            token_id: message.token_id(),
+            chain: message_direction.clone().into(),
+        }),
+    };
+
+    Ok(TokenChainInfo::new(
+        (message_direction, token_deployment_type).into(),
+    ))
```

Comment:
```suggestion
    match (message, token_chain_info) {
        (_, Some(token_chain_info)) => Ok(token_chain_info),
        (Message::InterchainTransfer { .. }, None) => bail!(Error::TokenNotDeployed {
            token_id: message.token_id(),
            chain: message_direction.clone().into(),
        }),
        (Message::DeployInterchainToken { minter: None, .. }, None) => Ok(TokenChainInfo::new(
            (message_direction, TokenDeploymentType::Trustless).into(),
        )),
        (Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. }, None) => Ok(
            TokenChainInfo::new((message_direction, TokenDeploymentType::CustomMinter).into()),
        ),
    }
```

Reply:
I didn't find that to be too readable so I separated the check